From cb8b5410ee4f6c4f3e96064a15e7dfea20b779b3 Mon Sep 17 00:00:00 2001
From: Li Nan <linan122@huawei.com>
Date: Thu, 23 Mar 2023 16:54:49 +0800
Subject: [PATCH 0070/1646] [Huawei] md/raid10: fix io hung in
 md_wait_for_blocked_rdev()

Offering: HULK
hulk inclusion
category: bugfix
bugzilla: 188569

--------------------------------

If badblocks are merged but bb->count exceedded, badblocks_set() will
return 1 and merged badblocks will become un-ack. rdev_set_badblocks()
will not set sb_flags and wakeup mddev->thread, io wait in
md_wait_for_blocked_rdev() will hung because BlockedBadBlocks may not be
cleared.

Fix it by checking badblocks->changed instead of return value. This flag
is set when badblocks changes.

Signed-off-by: Li Nan <linan122@huawei.com>
---
 drivers/md/md.c | 6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)

diff --git a/drivers/md/md.c b/drivers/md/md.c
index 306b294ba0b1..55a2800ab33e 100644
--- a/drivers/md/md.c
+++ b/drivers/md/md.c
@@ -9586,13 +9586,13 @@ int rdev_set_badblocks(struct md_rdev *rdev, sector_t s, int sectors,
 		       int is_new)
 {
 	struct mddev *mddev = rdev->mddev;
-	int rv;
+
 	if (is_new)
 		s += rdev->new_data_offset;
 	else
 		s += rdev->data_offset;
-	rv = badblocks_set(&rdev->badblocks, s, sectors, 0);
-	if (rv == 0) {
+	badblocks_set(&rdev->badblocks, s, sectors, 0);
+	if (rdev->badblocks.changed) {
 		/* Make sure they get written out promptly */
 		if (test_bit(ExternalBbl, &rdev->flags))
 			sysfs_notify_dirent_safe(rdev->sysfs_unack_badblocks);
-- 
2.17.1

